### DISCUSSION TITLE
how to create db driven botman?

### DISCUSSION THREAD

**Discussion by sherwinaval** [botman/botman#1387] [Q&A]
I wish to create something like this but should by dynamically loaded from the database

$bot->reply(Question::create('Please click from the options below to know more about our shop.')->addButtons([
                Button::create('Product Menu')->value('product_menu'),
                Button::create('How to Order')->value('how_to_order'),
                Button::create('Our Location')->value('our_location'),
                Button::create('Delivery Rates')->value('delivery_rates'),
                Button::create('Product Availability')->value('product_availability'),
                Button::create('Payment Modes')->value('payment_modes'),
            ]));

the Button::create part should be loaded from the database and must be looped. tried array but it wont work.

Someone help please?

**Comment by filippotoso** [comment]
Use a Conversation: https://botman.io/2.0/conversations

The should have a protected property with the current question ID from the DB (null at the beginning).

Create a method that uses the question ID to load the question from the database. If the question ID is null, load the first question.
In the method, use $this->ask() to ask your question and wait for an answer.

Within the ask() callback, store the answer in the database.
Then find the next ID from the DB. If it exists, set the property accordingly and call the same method again.
Otherwise (no more questions), call another method to complete the conversation.

**Comment by sherwinaval** [comment]
> Use a Conversation: https://botman.io/2.0/conversations
> 
> The should have a protected property with the current question ID from the DB (null at the beginning).
> 
> Create a method that uses the question ID to load the question from the database. If the question ID is null, load the first question. In the method, use $this->ask() to ask your question and wait for an answer.
> 
> Within the ask() callback, store the answer in the database. Then find the next ID from the DB. If it exists, set the property accordingly and call the same method again. Otherwise (no more questions), call another method to complete the conversation.

Thank you for responding, but I actually don't get what you mean... However let me show you what I tried doing...

```
$botman = resolve('botman');

$botman->hears('{message}', function ($bot, $message) {
	
	$server = 'localhost';
	$db = 'db';
	$dbuser = 'user';
	$dbpass = 'pass';

	//User Information
	$user = $bot->getUser();
	

	$conn = mysqli_connect($server, $dbuser, $dbpass, $db);
	
	if ($conn->connect_error) {
		die("Connection failed: " . $conn->connect_error);
	} 
		
	$sql = "SELECT id FROM tbl_receive WHERE message = '".strtolower($message)."'";
	$result = $conn->query($sql);
	$rcv_id = '';
	
	if ($result->num_rows > 0) {
		while ($row = $result->fetch_assoc()) {
			$rcv_id = $row["id"];
		}
	}
	
	if (!empty($rcv_id)) {
		$sql = "SELECT id, type, txt, value FROM tbl_response WHERE rcv_id = '".$rcv_id."' ORDER BY ord";
		$result = $conn->query($sql);
		
		if ($result->num_rows > 0) {
			while ($row = $result->fetch_assoc()) {
				switch($row['type']) {
					case 'text':
						$bot->reply($row['txt']);
						break;
					case 'image':
						$img = new Image($row['value']);
						$image = OutgoingMessage::create($row['txt'])->withAttachment($img);
						$bot->reply($image);
						break;
					case 'question':
						$question = 
							Question::create($row['txt'])
								->addButtons(
									[
										function() {
											$qsql = 'SELECT * FROM tbl_question_buttons WHERE resp_id = '.$row['id'].' ORDER BY ord';
											$qresult = $conn->query($qsql);

											if ($qresult->num_rows > 0) {
						
												while ($qrow = $qresult->fetch_assoc()) {
													Button::create($qrow['txt'])->value($qrow['value']);
												}
											}
											
										}
									]
								)
						;
						$bot->reply($question);
						break;
					case 'generictemplate':
					
						$gsql = 'SELECT * FROM tbl_generictemplate_elements WHERE resp_id = '.$row['id'].' ORDER BY ord';
						$gresult = $conn->query($gsql);
						
						if ($gresult->num_rows > 0) {
							$generictemplate=
								GenericTemplate::create()
									->addImageAspectRatio(GenericTemplate::RATIO_SQUARE)
									->addElements([
										function() {
											while ($grow = $gresult->fetch_assoc()) {
												Element::create($grow['name'])
													->subtitle($grow['subtitle'])
													->image($grow['imgurl'])
													->addButton(ElementButton::create($grow['btn_txt'])
														->payload($grow['btn_payload'])
														->type($grow['btn_type'])
												);
											}
										}
									]
								)
							;							
						}
						
						$bot->reply($generictemplate);
						break;
				}
				
			}
		}
	}
	
	$conn->close();
});
```

The text and image works without an issue. The generictemplate isn't fully tested but it works without an error. The question however throws an error.

[2023-11-02 12:31:40] local.ERROR: Call to undefined method Closure::toArray() {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Call to undefined method Closure::toArray() at \\vendor\\botman\\botman\\src\\Messages\\Outgoing\\Question.php:94)
[stacktrace]

**Comment by filippotoso** [comment]
Use a Convesation class.

**Comment by sherwinaval** [comment]
im fairly new to this  kind of thing.

But i'll try and follow your finger. Thank you
